### ISSUE TITLE
Feature Request: IMAP/POP3 for Statistics + Others

### ISSUE THREAD

**Issue by mubix** [gophish/gophish#129]
It would be nice to have the ability to configure an IMAP target so that if a response is received. In the future possible even parsing the response for email headers (IPs, hostnames, AV, etc) as well as identifying when responses are "Out of Office" responses, bounces, etc.

I've found that in phishing, these responses are gold not only for statistics , but also for re configuring items for the next phish.

**Comment by jordan-wright** [comment]
Interesting idea - thanks! I'll be thinking about how we can implement this to make it happen.

**Comment by rfdevere** [comment]
Sometimes I have received email responses that have sent me to fits of laughter, for this reason alone it should be a prominent feature. There is always the guy that emails you back and I've had a few send me their password in plain text in an email because my cloned site was blocked by their firewall.

**Comment by securitygeneration** [comment]
+1 on this request for me.

The ability to get out-of-office replies, but also elicit read receipts (which you can do with GoPhish by inserting `Disposition-Notification-To: Name <your@email.com>` as a custom header in the sending profile.

Some people have automatic responses to read receipts turned on (others simply click yes to read receipt prompts). Any received read receipts could obviously count towards the 'email opened' stat, even if they didn't load images.

Perhaps a config to parse emails received into an POP/IMAP mailbox? Most out of office and read receipt emails have a common (regexable) subject line.

The only way to do this currently is to run a separate daemon to monitor the mailbox (parse out the email address and find out their RId), and then make a call to the tracking image for that RId.

**Comment by stusmall** [comment]
I'm interested in taking this feature on as part of my day job.  How open are you to pull requests for this?  Also before I get started, do you have any architectural concerns or notes on what you want this to look like?

**Comment by glennzw** [comment]
We've added a bunch of support for IMAP recently (for reporting emails) - that might be a good starting point.

**Comment by stusmall** [comment]
In looking at this, there is an assumption that there is a single IMAP profile per user baked into the API.  I was hoping to be able to create multiple IMAP profiles that are then tied to campaigns like how there are multiple sending profilers.

Is there a preference on this?  Is it better to break the IMAP API so it isn't a one to one relationship with users or create a new concept of IMAP profiles?

I'm sure I'll have more questions since I've just started digging into this.

**Comment by glennzw** [comment]
The existing IMAP code is to monitor an account for reported emails.

I suspect we'd want a different account(s) for monitoring replies / bounces. You could use the boiler plate `imap/imap.go` but create new tooling for handling bounces / replies.

Also the upcoming reported UI stuff might be of interest for rendering https://github.com/gophish/gophish/issues/1851

**Comment by stusmall** [comment]
That makes sense.  My first pass was trying to reuse the existing IMAP table and API but it seems like it will actually be easier for me to create a new table and API for receiving profiles for campaign inboxes.  The only problem I see is that the API naming might be confusing since /api/imap is used for reporting.

I could either:
1.  Break the API for /api/imap and change it to /api/reporting-imap or something similar.  I looked and it doesn't seem to be documented yet so that shouldn't break compatibility for user.  I can then use /api/imap for setting up campaign receiving profiles.
2.  Just create an endpoint named something like /api/campaign-imap and do all my work there.

I don't have any strong opinions but want to make sure what I'm doing fits in with y'all's API design.

Thanks for the link to the other PR.  That looks like useful reference for when I start the UI work.

**Comment by glennzw** [comment]
I don't have any strong feelings either way but @jordan-wright has a good sense for idiomatic naming.

**Comment by securitygeneration** [comment]
Don’t want to hijack the thread, but I read that as “idiotic naming” and thought that was rather harsh 😛

**Comment by stusmall** [comment]
Is looking up a result by campaign id and email sufficiently unique?  From everything I see in the code it looks like it.  Pulling the rid from the DSN will require a non-negligible amount of work and it is hard to justify it when I already have campaign id and email in the same context.

**Comment by jordan-wright** [comment]
@stusmall Yes, each recipient (by email) in the campaign is unique.

I gotta be honest- I'm still a bit on the fence as to whether or not I'm interested in adding this to Gophish. It feels like a substantial amount of complexity, not only in the backend but also trying to reliably display this information in a user friendly way.

For example, let's say that we have an IMAP account per campaign. For every IMAP account created, Gophish will need to manage a background goroutine to poll that account. This isn't too terrible now since we limit it to one account per user, but this complexity will definitely be something we'd want to work out.

I also agree with the confusion between the existing IMAP functionality and the per-campaign IMAP functionality. For example, if I'm setting up Gophish and I configure an IMAP account, wouldn't I expect that reports that go to that account are marked as "reported"? How does that compare with the global IMAP account I've configured?

Personally, I'd almost prefer to focus on making #1851 happen, since this gives a huge net-add to Gophish's functionality by allowing it to be used as a triage utility. And maybe there are workarounds that we can use in the meantime just by making our existing IMAP functionality a bit smarter. For example (just thinking out loud here) would it be feasible to just setup new IMAP accounts with an auto forward rule to your global IMAP account and then have Gophish parse out those reports?

There's definitely a lot to think through here, but I appreciate all the great discussion!
